lib/rex/proto/kerberos/pac/krb5_pac.rb

```diff
@@ -589,20 +603,42 @@ def assign(val)
     end
 
     # Calculates the checksums, can only be done after all other fields are set
-    def calculate_checksums!(key: nil)
+    # @param [String] key Service key to calculate the server checksum
+    # @param [String] krbtgt_key key to calculate priv server (KDC) and full checksums with, if not specified `key` is used
+    def calculate_checksums!(key: nil, krbtgt_key: nil)
       server_checksum = nil
       priv_server_checksum = nil
+      full_pac_checksum = nil
       pac_info_buffers.each do |info_buffer|
         pac_element = info_buffer.buffer.pac_element
-        if pac_element.ul_type == 6
+        case pac_element.ul_type
+        when Krb5PacElementType::SERVER_CHECKSUM
           server_checksum = pac_element
-        elsif pac_element.ul_type == 7
+        when Krb5PacElementType::PRIVILEGE_SERVER_CHECKSUM
           priv_server_checksum = pac_element
+        when Krb5PacElementType::FULL_PAC_CHECKSUM
+          full_pac_checksum = pac_element
+        else
+          next
         end
       end
+
+      missing_checksums = []
+      missing_checksums << Krb5PacElementType::SERVER_CHECKSUM if server_checksum.nil?
+      missing_checksums << Krb5PacElementType::PRIVILEGE_SERVER_CHECKSUM if priv_server_checksum.nil?
+      raise Rex::Proto::Kerberos::Pac::Error::MissingInfoBuffer.new(ul_types: missing_checksums) unless missing_checksums.empty?
+
+      if krbtgt_key.nil?
+        krbtgt_key = key
+      end
+
+      if full_pac_checksum
```

Comment:
This logic will need to be updated in the future to include the `full_pac_checksum` element if it's required; At the minute the logic will only update an existing checksum, but it won't add a checksum element if it's meant to be added